From 7041545d50bbdeee9a9532b445479cb8446bb6fc Mon Sep 17 00:00:00 2001
From: Vinh Nguyen Trong <Vinh.NguyenTrong@vn.bosch.com>
Date: Mon, 22 Apr 2024 19:32:36 +0700
Subject: [PATCH 3/3] drm: rcar-du: cmm: check hardware buffer side of lut/clu
 setting

In the double buffering mode, RCAR-DU has two hardware buffers to
switch based on the vblank interrupts, in the meantime, the driver
accesses the other hardware buffer. The driver needs to update
both buffers in one setting. If the driver gets the same hardware
buffer at the second round of updating buffers, then the user will
encounter flickering.

Therefore, in this situation, the driver needs to skip writing data
to hardware buffer and re-trigger updating in the next vblank
interrupts.

Signed-off-by: Vinh Nguyen Trong <Vinh.NguyenTrong@vn.bosch.com>
Signed-off-by: Phong Hoang <phong.hoang.wz@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_cmm.c | 55 +++++++++++++++++++++++++--
 1 file changed, 52 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_cmm.c b/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
index 7c33dbc0f774..b0afe52fa42e 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
@@ -81,6 +81,7 @@
 #define CMM_LUT_NUM 256
 #define CMM_CLU_NUM (CMM_CLU_SAMPLES * CMM_CLU_SAMPLES * CMM_CLU_SAMPLES)
 #define CMM_HGO_NUM 64
+#define CMM_RETRY_NUM 30
 
 enum {
 	QUE_STAT_PENDING,
@@ -109,6 +110,8 @@ struct rcar_du_cmm_pending_event {
 
 struct cmm_module_t {
 	struct list_head list;
+	u32 retry;
+	u32 pre_addr;
 	union {
 		struct {
 			struct rcar_du_cmm_pending_event *p;
@@ -1022,10 +1025,10 @@ static inline void lc_event_done(struct cmm_module_t *module,
 
 	mutex_lock(&cmm_event_lock);
 
-	if (!done && list_empty(&module->list))
-		module->p = p;
-	else
+	if ((done || !list_empty(&module->list)) && !module->retry)
 		_event_done_locked(p);
+	else
+		module->p = p;
 
 	mutex_unlock(&cmm_event_lock);
 }
@@ -1131,6 +1134,40 @@ static int lut_table_copy(struct rcar_du_cmm *du_cmm)
 	return 0;
 }
 
+static bool is_same_buffer_side(struct rcar_du_cmm *du_cmm,
+				struct cmm_module_t *module,
+				u32 cur_addr)
+{
+	bool ret = false;
+	struct rcar_du_device *rcdu = du_cmm->rcrtc->group->dev;
+
+	if (!module->pre_addr) {
+		module->pre_addr = cur_addr;
+	} else if (module->pre_addr == cur_addr) {
+		module->retry++;
+
+		/* If the driver fails to reach the correct hardware buffer
+		 * after many times of retrying, the system is in a bad
+		 * situation. Give up reaching the correct hardware buffer
+		 * and update to the current buffer. The flickering on the
+		 * screen maybe observe.
+		 */
+		if (module->retry > CMM_RETRY_NUM) {
+			module->retry = 0;
+			module->pre_addr = 0;
+			dev_err(rcdu->dev,
+				"Failed to get HW buffer after %d retries\n",
+				CMM_RETRY_NUM);
+		} else {
+			ret = true;
+		}
+	} else {
+		module->retry = 0;
+		module->pre_addr = 0;
+	}
+	return ret;
+}
+
 /* set 1D look up table */
 static int lut_set(struct rcar_du_cmm *du_cmm,
 		   struct rcar_du_cmm_work_stat *stat)
@@ -1166,10 +1203,16 @@ static int lut_set(struct rcar_du_cmm *du_cmm,
 		return -EINVAL;
 	}
 
+	if (is_same_buffer_side(du_cmm, &du_cmm->lut, lut_base)) {
+		stat->done = false;
+		goto same_buffer_side;
+	}
+
 	lut_buf = gem_to_vaddr(stat->p->gem_obj);
 	for (i = 0; i < CMM_LUT_NUM; i++)
 		rcar_du_cmm_write(du_cmm, lut_base + i * 4, lut_buf[i]);
 
+same_buffer_side:
 	lc_event_done(&du_cmm->lut, stat->p, stat->done);
 
 	return 0;
@@ -1284,11 +1327,17 @@ static int clu_set(struct rcar_du_cmm *du_cmm,
 		return -EINVAL;
 	}
 
+	if (is_same_buffer_side(du_cmm, &du_cmm->clu, addr_reg)) {
+		stat->done = false;
+		goto same_buffer_side;
+	}
+
 	clu_buf = gem_to_vaddr(stat->p->gem_obj);
 	rcar_du_cmm_write(du_cmm, addr_reg, 0);
 	for (i = 0; i < CMM_CLU_NUM; i++)
 		rcar_du_cmm_write(du_cmm, data_reg, clu_buf[i]);
 
+same_buffer_side:
 	lc_event_done(&du_cmm->clu, stat->p, stat->done);
 
 	return 0;
-- 
2.25.1

